File: polyglot/Prototype/Prototype.java
// Prototype design pattern
// Imagine that in order to instantiate certain objects we need to carry out
// expensive database queries. In order to improve performance, it would 
// be beneficial to cache whatever object is created from the database
// and return a copy of the cached object whenever a subsequent request
// for object instantiation arises.

import java.util.*;

abstract class Shape implements Cloneable {

  private String id;

  abstract public void draw();


  public String getId(){
    return id;
  }

  public void setId(String id){
   this.id = id;
  }

  @Override
  public Object clone(){

    Object clone = null;

    try{
      clone = super.clone();
    } catch (CloneNotSupportedException e){
      e.printStackTrace();
    }

    return clone;
  }
}

class Rectangle extends Shape {

  @Override
  public void draw(){
    System.out.println("Inside Rectangle::draw() method.");
  }
}

class Square extends Shape {

  @Override
  public void draw(){
    System.out.println("Inside Square::draw() method.");
  }
}

class Circle extends Shape {

  @Override
  public void draw(){
    System.out.println("Inside Circle::draw() method.");
  }
}


class PrototypeManager {

  public PrototypeManager(){
    shapeMap = new Hashtable<String,Shape>();
  }

  public Shape getShape(String shapeId){
    Shape cachedShape = shapeMap.get(shapeId);
    return (Shape) cachedShape.clone();
  }

  // for each shape, run database query and create shape prototype
  // to be saved in cache
  public void loadCache(){

    Rectangle rect = new Rectangle();
    rect.setId("1");
    shapeMap.put(rect.getId(),rect);

    Square square = new Square();
    square.setId("2");
    shapeMap.put(square.getId(),square);

    Circle circle = new Circle();
    circle.setId("3");
    shapeMap.put(circle.getId(),circle);

  }

  private Hashtable<String,Shape> shapeMap;
}

public class Prototype {


  public static void main(String[] args){

    // need a prototype manager
    PrototypeManager manager = new PrototypeManager();
    // prototype manager registers a few prototype
    manager.loadCache();
    // prototype manager can now be used as factory object
    Shape clonedShape1 = manager.getShape("1");
    clonedShape1.draw();

    Shape clonedShape2 = manager.getShape("2");
    clonedShape2.draw();

    Shape clonedShape3 = manager.getShape("3");
    clonedShape3.draw();
  }

}

File: polyglot/Prototype/prototype.js
// Prototype design pattern
// Imagine that in order to instantiate certain objects we need to carry out
// expensive database queries. In order to improve performance, it would 
// be beneficial to cache whatever object is created from the database
// and return a copy of the cached object whenever a subsequent request
// for object instantiation arises.

function Shape(){
}
Shape.prototype.draw = function(){
  print("Shape::draw is not implemented");
}
Shape.prototype.getId = function(){
  return this.id;
}
Shape.prototype.setId = function(id){
  this.id = id;
}
Shape.prototype.clone = function(){
  print("Shape::clone is not implemented");
}

function Rectangle(){
  Shape.call(this);
}
Rectangle.prototype = Object.create(Shape.prototype);
Rectangle.prototype.draw = function(){
  print("Inside Rectangle::draw() method.");
}
Rectangle.prototype.clone = function(){
  clone = new Rectangle();
  clone.setId(this.id);
  return clone;
}

function Square(){
  Shape.call(this);
}
Square.prototype = Object.create(Shape.prototype);
Square.prototype.draw = function(){
  print("Inside Square::draw() method.");
}
Square.prototype.clone = function(){
  clone = new Square();
  clone.setId(this.id);
  return clone;
}

function Circle(){
  Shape.call(this);
}
Circle.prototype = Object.create(Shape.prototype);
Circle.prototype.draw = function(){
  print("Inside Circle::draw() method.");
}
Circle.prototype.clone = function(){
  clone = new Circle();
  clone.setId(this.id);
  return clone;
}

function PrototypeManager(){
  this.shapeMap = {}
}
PrototypeManager.prototype.loadCache = function(){
  rect = new Rectangle();
  rect.setId("1");
  this.shapeMap[rect.getId()] = rect;

  square = new Square();
  square.setId("2");
  this.shapeMap[square.getId()] = square;

  circle = new Circle();
  circle.setId("3");
  this.shapeMap[circle.getId()] = circle;
}
PrototypeManager.prototype.getShape= function(shapeId){
  cachedShape = this.shapeMap[shapeId];
  return cachedShape.clone();
}

// need a prototype manager
manager = new PrototypeManager();
// prototype manager registers a few prototypes
manager.loadCache();
// prototype manager can now be used as factory object
clonedShape1 = manager.getShape("1");
clonedShape1.draw();

clonedShape2 = manager.getShape("2");
clonedShape2.draw();

clonedShape3 = manager.getShape("3");
clonedShape3.draw();


